### ISSUE TITLE
[models]: slicing with changing compile-time-argument index

### ISSUE THREAD

**Issue by tchedaTT** [tenstorrent/tt-metal#43007]
### Component / Area

models

### Issue Type (optional)

Other

### Observed

Multiple models use a line similar to the following
https://github.com/tenstorrent/tt-metal/blob/2c1b1344ab3f0dbb077b6a0f4b8252798c6745b4/models/demos/llama3_70b_galaxy/tt/llama_model.py#L545
to eg. extract the last token's logits in prefill for sampling.
This gets compiled to a ttnn.slice with the index being a compile time argument.
As a consequence, whenever the true prefill length changes (other ops usually use a power-of-2-padded length), a new program compilation will happen.
This is not only wasteful, but can also technically lead to memory corruption if tracing is used (it is legal for an executing trace to overwrite any memory that was not allocated at the point when it was captured, and when a compilation happens it allocates new memory for its program cache buffer). The latter is however unlikely due to memory layout, unless at very high memory utilization.

### Expected

There is a version/overload  of ttnn.slice where the index is passed as tensor (NB it is NOT included in ttnn.slice docs). It should be used instead. When this is done, in some models the tracing scope can be extended to cover the slicing.

### 1. Steps (exact commands)

To verify whether a new compilation happens you can use the trace allocation tracker https://github.com/tenstorrent/tt-metal/tree/tcheda/trace-allocation-tracker

**Comment by tchedaTT** [comment]
TODO: also add a warning when tracing a slice variant with compile-time constant index
